hazardous/metrics/_accuracy_in_time.py

```diff
@@ -0,0 +1,142 @@
+
+    y_pred : array of shape (n_samples_test, n_events, n_time_grid)
+        Cumulative incidence for all competing events, at the time points
+        from the input time_grid.
+
+    time_grid : array of shape (n_time_grid,)
+        Time points used to predict the cumulative incidence.
+
+    quantiles : array or list of shape (n_quantiles,), default=None
+        The quantiles of ``time_grid`` used to define the fixed horizons at which
+        to compute the accuracy in time. The values of the quantiles of ``time_grid``
+        are equivalent to ``taus``, therefore ``quantiles`` can't be set if ``taus``
+        is set. If neither ``taus`` nor ``quantiles`` are set, we set quantiles
+        as a uniform grid of 8 quantiles, from 0 to 1.
+
+    taus : array or list of shape (n_taus), default=None
+        The fixed time horizons to compute the accuracy in time. Can't be set if
+        ``quantiles`` is set.
+
+    Returns
+    -------
+    acc_in_time : array of shape (n_quantiles or n_taus)
+        The accuracy in time computed at the fixed horizons ``taus``.
+
+    taus : array of shape (n_quantiles or n_taus)
+        The fixed time horizons effectively used to compute the accuracy in time.
+
+    References
+    ----------
+    .. [Alberge2024] J. Alberge, V. Maladiere,  O. Grisel, J. Abécassis, G. Varoquaux,
+        "Survival Models: Proper Scoring Rule and Stochastic Optimization
+        with Competing Risks", 2024
+    """
+    event_true, _ = check_y_survival(y_test)
+
+    if y_pred.ndim != 3:
+        raise ValueError(
+            "'y_pred' must be a 3D array with shape (n_samples, n_events, n_times), got"
+            f" shape {y_pred.shape}."
+        )
+    if y_pred.shape[0] != event_true.shape[0]:
+        raise ValueError(
+            "'y_true' and 'y_pred' must have the same number of samples, "
+            f"got {event_true.shape[0]} and {y_pred.shape[0]} respectively."
+        )
+    time_grid = np.atleast_1d(time_grid)
+    if y_pred.shape[2] != time_grid.shape[0]:
+        raise ValueError(
+            f"'time_grid' length ({time_grid.shape[0]}) "
+            f"must be equal to y_pred.shape[2] ({y_pred.shape[2]})."
+        )
+
+    if quantiles is not None:
+        if taus is not None:
+            raise ValueError("'quantiles' and 'taus' can't be set at the same time.")
+
+        quantiles = np.atleast_1d(quantiles)
+        if any(quantiles < 0) or any(quantiles > 1):
+            raise ValueError(f"quantiles must be in [0, 1], got {quantiles}.")
+        taus = np.quantile(time_grid, quantiles)
+
+    elif quantiles is None and taus is None:
+        n_quantiles = min(time_grid.shape[0], 8)
+        quantiles = np.linspace(0, 1, n_quantiles)
+        taus = np.quantile(time_grid, quantiles)
+
+    acc_in_time = []
+
+    for tau in taus:
+        mask_past_censored = (y_test["event"] == 0) & (y_test["duration"] <= tau)
+
+        tau_idx = np.searchsorted(time_grid, tau)
```

Comment:
time_grid needs to be sorted.

Reply:
Yes, although that's a common assumption, it might be good to sort it and sort the columns of y_pred? Or would you raise a warning?

Reply:
I added a warning, and a reordering with a copy